intellij/src/saros/intellij/ui/views/buttons/AbstractToolbarButton.java

```diff
@@ -1,33 +1,74 @@
 package saros.intellij.ui.views.buttons;
 
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Disposer;
 import com.intellij.ui.JBColor;
 import com.intellij.util.ui.UIUtil;
 import java.awt.Color;
+import javax.swing.AbstractButton;
 import javax.swing.BorderFactory;
 import javax.swing.ImageIcon;
 import javax.swing.JButton;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import saros.intellij.ui.views.SarosMainPanelView;
 
-/** Common class for Toolbar button implementations. */
-abstract class AbstractToolbarButton extends JButton {
+/**
+ * Common class for Toolbar button implementations.
+ *
+ * <p><b>NOTE:</b>This component and any component added here must be correctly torn down when the
+ * project the components belong to is closed. See {@link SarosMainPanelView}. This class offers the
+ * method {@link #dispose()} for such a purpose.
+ */
+abstract class AbstractToolbarButton extends JButton implements Disposable {
 
   // Self-adjusting references to the current IDE color
   static final Color FOREGROUND_COLOR = JBColor.foreground();
   static final Color BACKGROUND_COLOR = JBColor.background();
 
+  protected final Project project;
+
   /** Creates a button with the specified actionCommand, Icon and toolTipText. */
-  AbstractToolbarButton(String actionCommand, String tooltipText, ImageIcon icon) {
+  AbstractToolbarButton(
+      @NotNull Project project,
+      @Nullable String actionCommand,
+      @Nullable String tooltipText,
+      @Nullable ImageIcon icon) {
+
+    this.project = project;
+
+    Disposer.register(project, this);
+
     setActionCommand(actionCommand);
     setButtonIcon(icon);
     setToolTipText(tooltipText);
 
     setBorder(BorderFactory.createLineBorder(JBColor.border(), 1, true));
   }
 
-  /** calls {@link #setEnabled(boolean)} from the UI thread. */
-  void setEnabledFromUIThread(final boolean enabled) {
-    UIUtil.invokeAndWaitIfNeeded((Runnable) () -> setEnabled(enabled));
+  /**
+   * Calls {@link AbstractButton#setEnabled(boolean)} on the UI thread.
+   *
+   * @param enabled <code>true</code> to enable the button, <code>false</code> to disable the button
+   * @see UIUtil#invokeAndWaitIfNeeded(Runnable)
+   */
+  @Override
+  public void setEnabled(boolean enabled) {
+    UIUtil.invokeAndWaitIfNeeded((Runnable) () -> super.setEnabled(enabled));
```

Comment:
Using such syncExec calls gave us a lot of dead locks in Eclipse/E. I do not even know if it is good to coordinate the access to the widget here however I suggest to use ```invokeLaterIfNeeded``` instead.

Reply:
To be honest, I just made this change to be consistent. Both this and the default `setEnabled()` were used and I thought if someone bothered to add it there surely would be a need/purpose for it.

From reading the IntelliJ documentation on GUI interactions, I couldn't find any specification on whether to enable components synchronously. (And IntelliJ doesn't actively seem to mind as calls to the default `setEnabled()` work just fine.) But I am not well versed in the are of GUI programming. So if there are any other reasons to do this synchronously in a dedicated thread, please tell me.

Otherwise, I would simply adjust the corresponding commit by removing the `setEnabled()` implementation and instead using the default one.